### PR TITLE
Issue 1130 and 1131 smdb norc tools

### PR DESCRIPTION
**Pull request by Gbarra9**
This PR is related to 
- https://github.com/episphere/connect/issues/1130
- https://github.com/episphere/connect/issues/1131

- smdb code: https://github.com/episphere/studyManagerDashboard/pull/33

Reference Derived Variables Rules: https://nih.app.box.com/file/1668730770583?s=kc8ugxvosd9h5z3glinufwr0thvwlpgi

----
### Information of updates

**Reset Participant's SSN Survey**

        ssnStatusFlag (126331570) - 972455046 (not started num type)
        ssnSurveyStartTime (943232079) - delete
        ssnSurveyCompletedTime (315032037) - delete
        ssnFullGiven (311580100) - no
        ssnPartialGiven (914639140) - no
        ssnFullGivenTime (454067894) - delete
        ssnPartialGivenTime (598680838) - delete

**Participant Update Incentive Eligibility Update**

130371375.266600170.731498909 (Payment Round, baseline, Eligible for Payment) - 353358909 (Yes)

130371375.266600170.222373868 (Payment Round, baseline, NORC Payment Eligibility) = 353358909 (Yes)

130371375.266600170.787567527 (Payment Round, baseline, auto generated date when payment eligible for payment round) -<Current ISO 8601 Timestamp>

----

Code changes:

Added new endpoints for dashboard API resetParticipantSurvey, checkParticipantForEligibleIncentive, and updateParticipantIncentiveEligibility

### CODE REVIEW

File: utils/firestore.js [hunk 1]
```diff
@@ -3539,6 +3539,196 @@ const updateParticipantCorrection = async (participantData) => {
...
+        const siteCode = participantData[healthCareProvider];
+        const token = participantData['token'];
+        const specimenArray = await getSpecimenCollections(token, siteCode);
+
+        let incentiveEligible = false;
+
+        const currentPaymentRoundName = currentPaymentRound; // baseline or future payment rounds
+
+        // Add option to handle future payment rounds here later - left paymentRound as a parameter for now
+        if (participantData[paymentRound][currentPaymentRoundName][eligibleForIncentive] === no) {
```
Comment (amber-emmes) [inline utils/firestore.js:3634]: Is there any possibility that any of these objects will be null/undefined?

File: utils/firestore.js [hunk 2]
```diff
@@ -3539,6 +3539,196 @@ const updateParticipantCorrection = async (participantData) => {
...
+            
+            if (module1 && module2 && module3 && module4) {
+                if(bloodCollected) {
+                    incentiveEligible = true;
+                }    
+                else {
+                    if (specimenArray.length > 0) {
+                        const baselineResearchCollections = specimenArray.filter(collection => collection[biospecimenVisit] === baseline && collection[collectionSetting] === researchCollectionSetting);
+                    
+                        if (baselineResearchCollections.length != 0) {
```
Comment (amber-emmes) [inline utils/firestore.js:3650]: Probably better as !==

File: utils/firestore.js [hunk 3]
```diff
@@ -3539,6 +3539,196 @@ const updateParticipantCorrection = async (participantData) => {
...
+                }
+            }
+        }
+        return { 
+            isEligibleForIncentive: incentiveEligible, 
+            participantData 
+        };
+    } catch (error) {
+        console.error('Error checking participant for eligible incentive:', error);
+        throw {
```
Comment (amber-emmes) [inline utils/firestore.js:3672]: Your non-400 code error handlers use error.toString(), which I believe will result in '[object Object]'. This mismatch should probably be addressed.

File: utils/firestore.js [hunk 4]
```diff
@@ -3539,6 +3539,196 @@ const updateParticipantCorrection = async (participantData) => {
...
+
+        const eligibilityCheck = await checkParticipantForEligibleIncentive(connectId, currentPaymentRound);
+        const isEligibleForIncentive = eligibilityCheck.isEligibleForIncentive;
+        if (!isEligibleForIncentive) throw { code: 400, message: 'Participant is not eligible for incentive update.' };
+
+        const participantData = eligibilityCheck.participantData;
+        const currentPaymentRoundName = currentPaymentRound; // baseline or future payment rounds
+
+        const snapshot = await db.collection('participants').where('Connect_ID', '==', connectId).get();
+        if (snapshot.empty) throw { code: 404, message: 'Participant not found.' };
```
Comment (amber-emmes) [inline utils/firestore.js:3698]: Same comment about error.toString()

File: utils/firestore.js [hunk 5]
```diff
@@ -3539,6 +3539,196 @@ const updateParticipantCorrection = async (participantData) => {
...
+
+        if (participantData[paymentRound][currentPaymentRoundName][eligibleForIncentive] === no) {
+            console.log("🚀 ~ updateParticipantIncentiveEligibility ~ connectId, paymentRound, dateOfEligibility:", connectId, currentPaymentRound, dateOfEligibility)
+            await participantRef.update({
+                [`${paymentRound}.${currentPaymentRoundName}.${eligibleForIncentive}`]: yes,
+                [`${paymentRound}.${currentPaymentRoundName}.${norcPaymentEligibility}`]: yes,
+                [`${paymentRound}.${currentPaymentRoundName}.${timestampPaymentEligibilityForRound}`]: dateOfEligibility
+            });
+            const updatedDoc = await participantRef.get();
+            if (!updatedDoc.exists) throw { code: 404, message: 'Updated document not found.' }
```
Comment (amber-emmes) [inline utils/firestore.js:3710]: Multiple non-400 errors here with an error message set, but also just a `throw error`, which will result in inconsistent logging.

**Review by amber-emmes — COMMENTED** [review]
Mostly good, but a few comments.

**Review by amber-emmes — APPROVED** [review]
Looks good!

File: utils/dashboard.js [hunk 6]
```diff
@@ -141,8 +141,79 @@ const dashboard = async (req, res) => {
...
+                return res.status(err.code).json(getResponseJSON(err.message, err.code));
+            }
+            return res.status(500).json(getResponseJSON(err.message, 500));
+        }
+    } else if (api === `updateParticipantIncentiveEligibility`) {
+        if (req.method !== 'POST') {
+            return res.status(405).json(getResponseJSON('Only POST requests are accepted!', 405));
+        }
+        let body = req.body;
+        const { connectId, currentPaymentRound, dateOfEligibility } = body;
```
Comment (anthonypetersen) [inline utils/dashboard.js:200]: does dateOfEligibility need to be null checked?

File: utils/firestore.js [hunk 7]
```diff
@@ -3539,6 +3539,186 @@ const updateParticipantCorrection = async (participantData) => {
...
+};
+
+/**
+ * Used for NORC Incentive Eligibility tool
+ * Check if participant is eligible for incentive using modified snippet of code from checkDerivedVariables
+ * @param {string} connectId - Connect ID of the participant
+ * @param {string} paymentRound - Payment round to check eligibility for participant
+ * @returns {object} - Object with isEligibleForIncentive boolean and participantData object
+*/
+const checkParticipantForEligibleIncentive = async (connectId, currentPaymentRound) => {
```
Comment (anthonypetersen) [inline utils/firestore.js:3606]: I might be misunderstanding part of the workflow, but why is this end-point necessary?

File: utils/firestore.js [hunk 8]
```diff
@@ -3539,6 +3539,186 @@ const updateParticipantCorrection = async (participantData) => {
...
+    try {
```
Comment (anthonypetersen) [inline utils/firestore.js:3607]: If you wanted to try a simpler approach, you might be able to use the existing function that checks for incentive eligible participants (called from incentive.js) and then see if the returned object contains the participant being checked.

**Review by anthonypetersen — APPROVED** [review]
everything looks fine minus the few comments - let's chat when you're back in the office

File: utils/firestore.js (hunk 7 above)
Comment (Gbarra9) [reply utils/firestore.js:3868]: This endpoint is used when a participant selects a current payment round from dropdown on the front end. The function checks for the derived variables that would consider someone incentive eligible. The function is used for the `checkParticipantForEligibleIncentive ` endpoint and `updateParticipantIncentiveEligibility` endpoint. I reused the `checkParticipantForEligibleIncentive` function inside the `updateParticipantEligibility` function as an extra check in conjunction with checking that the participant's NORC Payment Eligibility is 'No'.

**Review by amber-emmes — APPROVED** [review]
Looks good.

File: utils/firestore.js [hunk 9]
```diff
@@ -3943,18 +3869,17 @@ const updateParticipantIncentiveEligibility = async (connectId, currentPaymentRo
...
-        const currentPaymentRoundName = currentPaymentRound; // baseline or future payment rounds
-
         const snapshot = await db.collection('participants').where('Connect_ID', '==', connectId).get();
         if (snapshot.empty) throw { message: 'Participant not found.', code: 404 };
-
         const participantRef = snapshot.docs[0].ref;
+        const participantData = snapshot.docs[0].data();
+        const currentPaymentRoundName = currentPaymentRound; // baseline or future payment rounds
+
+        const isNORCPaymentEligible = participantData?.[paymentRound]?.[currentPaymentRound]?.[norcPaymentEligibility] === no;
```
Comment (anthonypetersen) [inline utils/firestore.js:3878]: not needed - the front end should be checking this condition... all the backend should need to do is verify that both pieces of required information are present, the round and the date

Comment (anthonypetersen) [reply utils/firestore.js:3878]: @Gbarra9 - I would update the logic to being both of those variables equaling `no`, we never update just one of them

**Review by anthonypetersen — APPROVED** [review]

File: utils/firestore.js (hunk 9 above)
Comment (Gbarra9) [reply utils/firestore.js:3878]: @anthonypetersen and @amber-emmes, thanks for the input an re-review.

`if (isNORCPaymentEligible && isIncentiveEligible) {...}`

**Pull request merged**